File: apps/order-service/internal/domain/order/order_item.go
package order

import "errors"

type OrderItem struct {
	productID string
	quantity  int
	unitPrice Money
}

func NewOrderItem(productID string, quantity int, unitPriceCents int64, currency string) (OrderItem, error) {
	if productID == "" {
		return OrderItem{}, errors.New("productID is required")
	}
	if quantity <= 0 {
		return OrderItem{}, errors.New("quantity must be positive")
	}
	price, err := NewMoney(unitPriceCents, currency)
	if err != nil {
		return OrderItem{}, err
	}
	return OrderItem{productID: productID, quantity: quantity, unitPrice: price}, nil
}

func (i OrderItem) ProductID() string { return i.productID }
func (i OrderItem) Quantity() int     { return i.quantity }
func (i OrderItem) UnitPrice() Money  { return i.unitPrice }
func (i OrderItem) Subtotal() (Money, error) { return i.unitPrice.Multiply(int64(i.quantity)) }

File: apps/domain-service/src/domain/order/value-objects/OrderItem.ts
import { ValueObject } from '@teste-manuel/domain';

export interface OrderItemProps {
  productId: string;
  quantity: number;
  unitPrice: number;
}

export class OrderItem extends ValueObject<OrderItemProps> {
  constructor(productId: string, quantity: number, unitPrice: number) {
    if (!productId) {
      throw new Error('Product ID is required');
    }
    if (!Number.isInteger(quantity) || quantity <= 0) {
      throw new Error('Quantity must be a positive integer');
    }
    if (unitPrice < 0) {
      throw new Error('Unit price cannot be negative');
    }
    super({ productId, quantity, unitPrice });
  }

  static create(productId: string, quantity: number, unitPrice: number): OrderItem {
    return new OrderItem(productId, quantity, unitPrice);
  }

  get productId(): string {
    return this.props.productId;
  }

  get quantity(): number {
    return this.props.quantity;
  }

  get unitPrice(): number {
    return this.props.unitPrice;
  }

  get subtotal(): number {
    return Math.round(this.props.quantity * this.props.unitPrice * 100) / 100;
  }

  equals(other: OrderItem): boolean {
    return (
      this.props.productId === other.props.productId &&
      this.props.quantity === other.props.quantity &&
      this.props.unitPrice === other.props.unitPrice
    );
  }
}
